File: FrigoTab.AcceptanceTests/T20260914T211700Z_001_RealSwitcherAcceptanceTests.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Threading;
using System.Windows.Forms;
using FrigoTab.Core;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace FrigoTab.AcceptanceTests {

    [STATestClass]
    [DoNotParallelize]
    [TestCategory("Acceptance")]
    public class T20260914T211700Z_001_RealSwitcherAcceptanceTests {

        [TestMethod]
        public void AltTabOpensTheRealSwitcherAndSelectsARealPreview () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();

                Assert.IsTrue(desktop.Form.Visible);
                Assert.AreEqual(SwitcherState.Visible, desktop.Switcher.State);
                Assert.AreEqual(desktop.Switcher.CandidateCount, desktop.Tiles.Count);
                Assert.AreEqual(0, desktop.Switcher.SelectedIndex);
                Assert.AreEqual(1, desktop.Tiles.Count(IsSelected));
            }
        }

        [TestMethod]
        public void HoldingAltCyclesForwardAndWrapsThroughRealPreviews () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();
                int count = desktop.Switcher.CandidateCount;
                Assert.IsTrue(count >= 3);

                for( int index = 1; index < count; index++ ) {
                    Assert.AreEqual(KeyHandling.Consume, desktop.Key(SwitcherKey.Tab, KeyTransition.Down, alt: true));
                    Assert.AreEqual(index, desktop.Switcher.SelectedIndex);
                }
                Assert.AreEqual(KeyHandling.Consume, desktop.Key(SwitcherKey.Tab, KeyTransition.Down, alt: true));
                Assert.AreEqual(0, desktop.Switcher.SelectedIndex);
            }
        }

        [TestMethod]
        public void ShiftAltTabCyclesBackwardAndWrapsThroughRealPreviews () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();
                int last = desktop.Switcher.CandidateCount - 1;

                Assert.AreEqual(KeyHandling.Consume, desktop.Key(SwitcherKey.Tab, KeyTransition.Down, alt: true, shift: true));
                Assert.AreEqual(last, desktop.Switcher.SelectedIndex);

                for( int index = last - 1; index >= 0; index-- ) {
                    desktop.Key(SwitcherKey.Tab, KeyTransition.Down, alt: true, shift: true);
                    Assert.AreEqual(index, desktop.Switcher.SelectedIndex);
                }
            }
        }

        [TestMethod]
        public void ReleasingAltLeavesTheRealSwitcherOpenForDeliberateSelection () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();
                ApplicationWindow target = desktop.FixtureTiles.First(tile => !IsSelected(tile));
                desktop.Switcher.HandleMouseMove(desktop.Center(target));
                Assert.AreSame(target, desktop.SelectedTile);

                Assert.AreEqual(KeyHandling.PassThrough, desktop.Key(SwitcherKey.Alt, KeyTransition.Up));

                Assert.AreEqual(SwitcherState.Visible, desktop.Switcher.State);
                Assert.IsTrue(desktop.Form.Visible);
                Assert.AreSame(target, desktop.SelectedTile);

                desktop.Switcher.HandleMouseClick(desktop.Center(target));
                desktop.Pump();
                desktop.AssertClosed();
                desktop.AssertForeground(target.Application);
            }
        }

        [TestMethod]
        public void EscapeCancelsTheRealSession () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();

                Assert.AreEqual(KeyHandling.Consume, desktop.Key(SwitcherKey.Escape, KeyTransition.Down));

                desktop.AssertClosed();
            }
        }

        [TestMethod]
        public void AltF4CancelsTheRealSession () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();

                Assert.AreEqual(KeyHandling.Consume, desktop.Key(SwitcherKey.F4, KeyTransition.Down, alt: true));

                desktop.AssertClosed();
            }
        }

        [TestMethod]
        public void NumberOneActivatesTheFirstRealPreview () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();
                WindowHandle expected = desktop.SelectedTile.Application;

                Assert.AreEqual(KeyHandling.Consume, desktop.Key(SwitcherKey.D1, KeyTransition.Down));

                desktop.AssertClosed();
                desktop.AssertForeground(expected);
            }
        }

        [TestMethod]
        public void NumberPadTwoActivatesTheSecondRealPreview () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();
                desktop.Key(SwitcherKey.Tab, KeyTransition.Down, alt: true);
                WindowHandle expected = desktop.SelectedTile.Application;
                desktop.Key(SwitcherKey.Tab, KeyTransition.Down, alt: true, shift: true);

                Assert.AreEqual(KeyHandling.Consume, desktop.Key(SwitcherKey.NumPad2, KeyTransition.Down));

                desktop.AssertClosed();
                desktop.AssertForeground(expected);
            }
        }

        [TestMethod]
        public void PointerHoverAndClickSelectAndActivateTheExactRealPreview () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();
                ApplicationWindow target = desktop.FixtureTiles.First(tile => !IsSelected(tile));
                ScreenPoint point = desktop.Center(target);

                desktop.Switcher.HandleMouseMove(point);
                Assert.AreSame(target, desktop.SelectedTile);
                desktop.Switcher.HandleMouseClick(point);
                desktop.Pump();

                desktop.AssertClosed();
                desktop.AssertForeground(target.Application);
            }
        }

        [TestMethod]
        public void MovingOutsideTheTilesClearsSelectionAndAltTabRestoresIt () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();

                desktop.Switcher.HandleMouseMove(desktop.PointOutsideTiles());
                Assert.IsNull(desktop.Switcher.SelectedIndex);
                Assert.AreEqual(0, desktop.Tiles.Count(IsSelected));

                desktop.Key(SwitcherKey.Tab, KeyTransition.Down, alt: true);
                Assert.AreEqual(0, desktop.Switcher.SelectedIndex);
                Assert.AreEqual(1, desktop.Tiles.Count(IsSelected));
            }
        }

        [TestMethod]
        public void AClosedTargetLeavesTheRealSessionAvailableForCancellation () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();
                ApplicationWindow target = desktop.FixtureTiles.First();
                desktop.Switcher.HandleMouseMove(desktop.Center(target));
                desktop.CloseFixture(target.Application);

                desktop.Key(SwitcherKey.Alt, KeyTransition.Up);

                Assert.AreEqual(SwitcherState.Visible, desktop.Switcher.State);
                Assert.IsTrue(desktop.Form.Visible);
                desktop.Key(SwitcherKey.Escape, KeyTransition.Down);
                desktop.AssertClosed();
            }
        }

        [TestMethod]
        public void ClosingAndReopeningRecreatesRealNativePreviews () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();
                int count = desktop.Tiles.Count;
                desktop.Key(SwitcherKey.Escape, KeyTransition.Down);
                desktop.AssertClosed();

                desktop.Open();

                Assert.AreEqual(count, desktop.Tiles.Count);
                Assert.AreEqual(1, desktop.Tiles.Count(IsSelected));
            }
        }

        [TestMethod]
        public void InjectedAltTabPassesThroughWithoutOpeningRealWindows () {
            using( LiveSession desktop = new LiveSession() ) {
                Assert.AreEqual(
                    KeyHandling.PassThrough,
                    desktop.Key(SwitcherKey.Tab, KeyTransition.Down, alt: true, injected: true));
                Assert.AreEqual(SwitcherState.Idle, desktop.Switcher.State);
                Assert.IsFalse(desktop.Form.Visible);
                Assert.AreEqual(0, desktop.Tiles.Count);
            }
        }

        [TestMethod]
        public void ConsumedKeyUpsRemainBalancedAfterTheRealSessionCloses () {
            using( LiveSession desktop = new LiveSession() ) {
                desktop.Open();
                desktop.Key(SwitcherKey.Escape, KeyTransition.Down);
                desktop.AssertClosed();

                Assert.AreEqual(KeyHandling.Consume, desktop.Key(SwitcherKey.Tab, KeyTransition.Up));
                Assert.AreEqual(KeyHandling.Consume, desktop.Key(SwitcherKey.Escape, KeyTransition.Up));
                Assert.AreEqual(KeyHandling.PassThrough, desktop.Key(SwitcherKey.Tab, KeyTransition.Up));
            }
        }

        [TestMethod]
        public void RepeatedOpenAndCloseLeavesNoPreviewFormsBehind () {
            using( LiveSession desktop = new LiveSession() ) {
                for( int repetition = 0; repetition < 5; repetition++ ) {
                    desktop.Open();
                    Assert.IsTrue(desktop.Tiles.Count > 0);
                    desktop.Key(SwitcherKey.Escape, KeyTransition.Down);
                    desktop.AssertClosed();
                    Assert.AreEqual(0, desktop.Tiles.Count);
                }
            }
        }

        private static bool IsSelected (ApplicationWindow tile) {
            Property<bool> selected = tile.Selected;
            return selected.Value;
        }

        private sealed class LiveSession : IDisposable {

            private readonly List<Form> fixtures = new List<Form>();
            private readonly Dictionary<WindowHandle, Form> fixturesByHandle = new Dictionary<WindowHandle, Form>();
            private readonly WindowHandle previousForeground;

            public LiveSession () {
                previousForeground = WindowHandle.GetForegroundWindow();
                Form = new SessionForm();
                _ = Form.Handle;

                Color[] colors = {Color.Magenta, Color.Lime, Color.Orange};
                for( int index = 0; index < colors.Length; index++ ) {
                    Form fixture = new Form {
                        Text = "FrigoTab real acceptance window " + (index + 1),
                        BackColor = colors[index],
                        Bounds = new Rectangle(80 + index * 90, 80 + index * 70, 640, 480),
                        StartPosition = FormStartPosition.Manual,
                        ShowInTaskbar = true
                    };
                    fixture.Show();
                    fixture.Refresh();
                    fixtures.Add(fixture);
                    fixturesByHandle.Add(new WindowHandle(fixture.Handle), fixture);
                }
                Pump();
                Switcher = new SwitcherApplication((ISwitcherSessionPort) Form);
            }

            public SessionForm Form { get; }
            public SwitcherApplication Switcher { get; }

            public List<ApplicationWindow> Tiles => Application.OpenForms
                .Cast<Form>()
                .OfType<ApplicationWindow>()
                .Where(tile => ReferenceEquals(tile.Owner, Form))
                .ToList();

            public List<ApplicationWindow> FixtureTiles => Tiles
                .Where(tile => fixturesByHandle.ContainsKey(tile.Application))
                .ToList();

            public ApplicationWindow SelectedTile => Tiles.Single(IsSelected);

            public void Open () {
                Assert.AreEqual(KeyHandling.Consume, Key(SwitcherKey.Tab, KeyTransition.Down, alt: true));
                WaitUntil(() => Form.Visible && Tiles.Count > 0);
                Assert.AreEqual(SwitcherState.Visible, Switcher.State);
            }

            public KeyHandling Key (
                SwitcherKey key,
                KeyTransition transition,
                bool alt = false,
                bool shift = false,
                bool injected = false) {
                KeyHandling result = Switcher.HandleKeyboard(new KeyboardInput(key, transition, alt, shift, injected));
                Pump();
                return result;
            }

            public ScreenPoint Center (ApplicationWindow tile) => new ScreenPoint(
                tile.Bounds.Left + tile.Bounds.Width / 2,
                tile.Bounds.Top + tile.Bounds.Height / 2);

            public ScreenPoint PointOutsideTiles () {
                Rectangle desktop = SystemInformation.VirtualScreen;
                for( int y = desktop.Top; y < desktop.Bottom; y += 8 ) {
                    for( int x = desktop.Left; x < desktop.Right; x += 8 ) {
                        Point point = new Point(x, y);
                        if( Tiles.All(tile => !tile.Bounds.Contains(point)) ) {
                            return new ScreenPoint(x, y);
                        }
                    }
                }
                Assert.Fail("No point outside the preview windows was available.");
                return default(ScreenPoint);
            }

            public void CloseFixture (WindowHandle handle) {
                Form fixture = fixturesByHandle[handle];
                fixturesByHandle.Remove(handle);
                fixture.Close();
                fixture.Dispose();
                fixtures.Remove(fixture);
                Pump();
            }

            public void AssertClosed () {
                WaitUntil(() => Switcher.State == SwitcherState.Idle && !Form.Visible);
            }

            public void AssertForeground (WindowHandle expected) {
                WaitUntil(() => WindowHandle.GetForegroundWindow() == expected);
            }

            public void Pump () {
                Application.DoEvents();
                Thread.Sleep(20);
                Application.DoEvents();
            }

            public void Dispose () {
                Switcher.Close();
                Form.Dispose();
                foreach( Form fixture in fixtures ) {
                    fixture.Close();
                    fixture.Dispose();
                }
                fixtures.Clear();
                fixturesByHandle.Clear();
                if( previousForeground != WindowHandle.Null ) {
                    previousForeground.SetForeground();
                }
                Pump();
            }

            private void WaitUntil (Func<bool> condition) {
                Stopwatch timeout = Stopwatch.StartNew();
                while( timeout.Elapsed < TimeSpan.FromSeconds(5) ) {
                    if( condition() ) {
                        return;
                    }
                    Pump();
                }
                Assert.Fail("Timed out waiting for the real desktop state.");
            }

        }

    }

}

File: acceptance-tests/tests/t20260914t211700z_001_real_switcher_acceptance_tests.rs
#![cfg(windows)]

use std::time::Duration;

use frigotab::key_handling::KeyHandling;
use frigotab::keyboard_input::{KeyTransition, KeyboardInput, SwitcherKey};
use frigotab::screen_point::ScreenPoint;
use frigotab::switcher_state::SwitcherState;
use frigotab_acceptance::{
    LiveSession, LiveTile, foreground_window, is_layered, is_window_visible, serial_guard,
    visible_owned_layered_windows, wait_until, window_bounds, window_owner,
};
use windows_sys::Win32::Foundation::{HWND, RECT};

const TIMEOUT: Duration = Duration::from_secs(5);

fn open_session(session: &mut LiveSession) {
    let result = session.open();
    assert_eq!(KeyHandling::Consume, result);
    assert!(
        session.wait_visible(TIMEOUT),
        "the real session did not become visible"
    );
    assert!(
        !session.tiles().is_empty(),
        "the real session has no preview HWNDs"
    );
    assert_eq!(SwitcherState::Visible, session.switcher_state());
    assert_eq!(session.candidate_count(), session.tiles().len());
    assert_eq!(Some(0), session.selected_index());
}

fn key(
    session: &mut LiveSession,
    key: SwitcherKey,
    transition: KeyTransition,
    alt: bool,
    shift: bool,
    injected: bool,
) -> KeyHandling {
    session.key(KeyboardInput::new(key, transition, alt, shift, injected))
}

fn center(tile: LiveTile) -> ScreenPoint {
    ScreenPoint::new(
        tile.bounds.left + (tile.bounds.right - tile.bounds.left) / 2,
        tile.bounds.top + (tile.bounds.bottom - tile.bounds.top) / 2,
    )
}

fn fixture_tiles(session: &LiveSession) -> Vec<LiveTile> {
    let fixtures = session.fixture_handles();
    session
        .tiles()
        .into_iter()
        .filter(|tile| fixtures.contains(&tile.source))
        .collect()
}

fn point_outside_tiles(tiles: &[LiveTile], bounds: RECT) -> ScreenPoint {
    for y in (bounds.top..bounds.bottom).step_by(8) {
        for x in (bounds.left..bounds.right).step_by(8) {
            if tiles.iter().all(|tile| {
                x < tile.bounds.left
                    || x >= tile.bounds.right
                    || y < tile.bounds.top
                    || y >= tile.bounds.bottom
            }) {
                return ScreenPoint::new(x, y);
            }
        }
    }
    panic!("no screen point outside the real preview HWNDs was available");
}

fn wait_foreground(expected: HWND) {
    assert!(
        wait_until(TIMEOUT, || foreground_window() == expected),
        "the selected real target did not become foreground"
    );
}

fn assert_hidden(session: &mut LiveSession) {
    assert!(
        wait_until(TIMEOUT, || {
            session.switcher_state() == SwitcherState::Idle && !is_window_visible(session.owner())
        }),
        "the real session did not close"
    );
    assert!(
        visible_owned_layered_windows(session.owner()).is_empty(),
        "a preview HWND remained visible after closing the session"
    );
    assert_eq!(SwitcherState::Idle, session.switcher_state());
    assert_eq!(0, session.candidate_count());
    assert_eq!(None, session.selected_index());
}

#[test]
fn alt_tab_opens_the_real_switcher_and_selects_a_real_preview() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();

    open_session(&mut session);

    assert!(is_window_visible(session.owner()));
    assert!(session.selected_tile().is_some());
    assert_eq!(1, session.selected_count());
    assert!(fixture_tiles(&session).len() >= 3);

    let tiles = session.tiles();
    let overlays = visible_owned_layered_windows(session.owner());
    assert_eq!(tiles.len(), overlays.len());
    for tile in tiles {
        assert!(is_window_visible(tile.popup));
        assert_eq!(session.owner(), window_owner(tile.popup));
        assert!(is_layered(tile.popup));
    }
}

#[test]
fn holding_alt_cycles_forward_and_wraps_through_real_previews() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();
    open_session(&mut session);

    let sources: Vec<HWND> = session
        .tiles()
        .into_iter()
        .map(|tile| tile.source)
        .collect();
    assert!(sources.len() >= 3);
    assert_eq!(Some(sources[0]), session.selected_source());

    for expected in sources.iter().skip(1) {
        let index = sources
            .iter()
            .position(|source| source == expected)
            .expect("the selected source disappeared from the real tile list");
        assert_eq!(
            KeyHandling::Consume,
            key(
                &mut session,
                SwitcherKey::Tab,
                KeyTransition::Down,
                true,
                false,
                false
            )
        );
        assert_eq!(Some(*expected), session.selected_source());
        assert_eq!(Some(index), session.selected_index());
        assert!(session.is_visible());
    }
    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::Tab,
            KeyTransition::Down,
            true,
            false,
            false
        )
    );
    assert_eq!(Some(sources[0]), session.selected_source());
    assert_eq!(Some(0), session.selected_index());
}

#[test]
fn shift_alt_tab_cycles_backward_and_wraps_through_real_previews() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();
    open_session(&mut session);

    let sources: Vec<HWND> = session
        .tiles()
        .into_iter()
        .map(|tile| tile.source)
        .collect();
    assert!(sources.len() >= 3);
    let last = sources.len() - 1;

    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::Tab,
            KeyTransition::Down,
            true,
            true,
            false
        )
    );
    assert_eq!(Some(sources[last]), session.selected_source());
    assert_eq!(Some(last), session.selected_index());
    for (index, expected) in sources[..last].iter().enumerate().rev() {
        assert_eq!(
            KeyHandling::Consume,
            key(
                &mut session,
                SwitcherKey::Tab,
                KeyTransition::Down,
                true,
                true,
                false
            )
        );
        assert_eq!(Some(*expected), session.selected_source());
        assert_eq!(Some(index), session.selected_index());
    }
}

#[test]
fn releasing_alt_leaves_the_real_switcher_open_for_deliberate_selection() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();
    open_session(&mut session);

    let initial = session.selected_source();
    let (target_index, target) = session
        .tiles()
        .into_iter()
        .enumerate()
        .find(|(_, tile)| Some(tile.source) != initial)
        .expect("at least two real previews are required");
    session.mouse_move(center(target));
    assert_eq!(Some(target.source), session.selected_source());
    assert_eq!(Some(target_index), session.selected_index());

    assert_eq!(
        KeyHandling::PassThrough,
        key(
            &mut session,
            SwitcherKey::Alt,
            KeyTransition::Up,
            false,
            false,
            false
        )
    );
    assert!(session.is_visible());
    assert!(is_window_visible(session.owner()));
    assert_eq!(Some(target.source), session.selected_source());
    assert_eq!(Some(target_index), session.selected_index());

    session.mouse_click(center(target));
    assert_hidden(&mut session);
    wait_foreground(target.source);
}

#[test]
fn escape_cancels_the_real_session() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();
    open_session(&mut session);

    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::Escape,
            KeyTransition::Down,
            false,
            false,
            false
        )
    );
    assert_hidden(&mut session);
}

#[test]
fn alt_f4_cancels_the_real_session() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();
    open_session(&mut session);

    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::F4,
            KeyTransition::Down,
            true,
            false,
            false
        )
    );
    assert_hidden(&mut session);
}

#[test]
fn number_one_activates_the_first_real_preview() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();
    open_session(&mut session);
    let expected = session
        .selected_source()
        .expect("the first real preview was not selected");
    assert_eq!(Some(0), session.selected_index());

    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::D1,
            KeyTransition::Down,
            false,
            false,
            false
        )
    );
    assert_hidden(&mut session);
    wait_foreground(expected);
}

#[test]
fn number_pad_two_activates_the_second_real_preview() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();
    open_session(&mut session);
    let sources: Vec<HWND> = session
        .tiles()
        .into_iter()
        .map(|tile| tile.source)
        .collect();
    assert!(sources.len() >= 2);

    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::Tab,
            KeyTransition::Down,
            true,
            false,
            false
        )
    );
    let expected = session
        .selected_source()
        .expect("Tab did not select a second preview");
    assert_eq!(Some(sources[1]), Some(expected));
    assert_eq!(Some(1), session.selected_index());
    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::Tab,
            KeyTransition::Down,
            true,
            true,
            false
        )
    );
    assert_eq!(Some(sources[0]), session.selected_source());
    assert_eq!(Some(0), session.selected_index());

    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::NumPad2,
            KeyTransition::Down,
            false,
            false,
            false,
        )
    );
    assert_hidden(&mut session);
    wait_foreground(expected);
}

#[test]
fn pointer_hover_and_click_select_and_activate_the_exact_real_preview() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();
    open_session(&mut session);
    let initial = session.selected_source();
    let target = fixture_tiles(&session)
        .into_iter()
        .find(|tile| Some(tile.source) != initial)
        .expect("a second real fixture preview is required");

    session.mouse_move(center(target));
    assert_eq!(Some(target.source), session.selected_source());
    assert_eq!(
        Some(
            session
                .tiles()
                .iter()
                .position(|tile| tile.source == target.source)
                .expect("the hovered fixture tile disappeared")
        ),
        session.selected_index()
    );
    session.mouse_click(center(target));

    assert_hidden(&mut session);
    wait_foreground(target.source);
}

#[test]
fn moving_outside_the_tiles_clears_selection_and_alt_tab_restores_it() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();
    open_session(&mut session);
    let tiles = session.tiles();
    let first = session.selected_source();
    let outside = point_outside_tiles(&tiles, window_bounds(session.owner()).unwrap());

    session.mouse_move(outside);
    assert_eq!(None, session.selected_source());
    assert_eq!(None, session.selected_index());
    assert_eq!(0, session.selected_count());
    assert!(session.is_visible());
    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::Tab,
            KeyTransition::Down,
            true,
            false,
            false
        )
    );
    assert_eq!(first, session.selected_source());
    assert_eq!(Some(0), session.selected_index());
    assert_eq!(1, session.selected_count());
}

#[test]
fn a_closed_target_leaves_the_real_session_available_for_cancellation() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();
    open_session(&mut session);
    let target = fixture_tiles(&session)
        .into_iter()
        .next()
        .expect("a real fixture preview is required");

    session.mouse_move(center(target));
    assert_eq!(Some(target.source), session.selected_source());
    assert!(session.selected_index().is_some());
    assert!(session.close_fixture(target.source));
    assert_eq!(
        KeyHandling::PassThrough,
        key(
            &mut session,
            SwitcherKey::Alt,
            KeyTransition::Up,
            false,
            false,
            false
        )
    );
    assert!(session.is_visible());
    assert!(is_window_visible(session.owner()));
    assert!(session.selected_index().is_some());

    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::Escape,
            KeyTransition::Down,
            false,
            false,
            false
        )
    );
    assert_hidden(&mut session);
}

#[test]
fn closing_and_reopening_recreates_real_native_previews() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();
    open_session(&mut session);
    let count = session.tiles().len();
    assert!(count > 0);

    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::Escape,
            KeyTransition::Down,
            false,
            false,
            false
        )
    );
    assert_hidden(&mut session);
    assert!(session.tiles().is_empty());

    open_session(&mut session);
    assert_eq!(count, session.tiles().len());
    assert!(session.selected_tile().is_some());
    assert_eq!(Some(0), session.selected_index());
}

#[test]
fn injected_alt_tab_passes_through_without_opening_real_windows() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();

    assert_eq!(
        KeyHandling::PassThrough,
        key(
            &mut session,
            SwitcherKey::Tab,
            KeyTransition::Down,
            true,
            false,
            true
        )
    );
    assert!(!session.is_visible());
    assert!(!is_window_visible(session.owner()));
    assert!(session.tiles().is_empty());
    assert_eq!(SwitcherState::Idle, session.switcher_state());
    assert_eq!(0, session.candidate_count());
    assert_eq!(None, session.selected_index());
}

#[test]
fn consumed_key_ups_remain_balanced_after_the_real_session_closes() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();
    open_session(&mut session);
    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::Escape,
            KeyTransition::Down,
            false,
            false,
            false
        )
    );
    assert_hidden(&mut session);

    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::Tab,
            KeyTransition::Up,
            false,
            false,
            false
        )
    );
    assert_eq!(
        KeyHandling::Consume,
        key(
            &mut session,
            SwitcherKey::Escape,
            KeyTransition::Up,
            false,
            false,
            false
        )
    );
    assert_eq!(
        KeyHandling::PassThrough,
        key(
            &mut session,
            SwitcherKey::Tab,
            KeyTransition::Up,
            false,
            false,
            false
        )
    );
    assert!(!session.is_visible());
}

#[test]
fn repeated_open_and_close_leaves_no_preview_forms_behind() {
    let _serial = serial_guard();
    let mut session = LiveSession::new();

    for _ in 0..5 {
        open_session(&mut session);
        assert!(!session.tiles().is_empty());
        assert_eq!(
            KeyHandling::Consume,
            key(
                &mut session,
                SwitcherKey::Escape,
                KeyTransition::Down,
                false,
                false,
                false
            )
        );
        assert_hidden(&mut session);
        assert!(session.tiles().is_empty());
    }
}
